### PR TITLE
Pil nsimage support

### PR DESCRIPTION
**Pull request by typemytype**
add support for

```python
pilImageObject = saveImage("PIL")
nsImageImageObject = saveImage("NSImage")
```

#451

### CODE REVIEW

**Comment by typemytype** [comment]
I guess this PR is ready!

File: drawBot/context/__init__.py [hunk 1]
```diff
@@ -17,7 +18,9 @@
     BMPContext,
     MP4Context,
     ICNSContext,
-    PrintContext
+    PrintContext,
+    PILContext,
+    NSImageContext
```
Comment (justvanrossum) [inline drawBot/context/__init__.py:23]:
```suggestion
    NSImageContext,
```

File: drawBot/context/gifContext.py [hunk 2]
```diff
@@ -43,7 +44,13 @@ def _writeDataToFile(self, data, path, options):
             options["multipage"] = True
             tempPath = tempfile.mkstemp(suffix=".gif")[1]
 
-        inputPaths = super(GIFContext, self)._writeDataToFile(data, tempPath, options)
+        self._inputPaths = []
+        super(GIFContext, self)._writeDataToFile(data, tempPath, options)
```
Comment (justvanrossum) [inline drawBot/context/gifContext.py:48]:
```suggestion
        super()._writeDataToFile(data, tempPath, options)
```

File: drawBot/context/tools/pilTools.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import AppKit
+import io
+
+
+def pilToNSImage(pilImage):
+    buffer = io.BytesIO()
+    pilImage.save(buffer, pilImage.format)
```
Comment (justvanrossum) [inline drawBot/context/tools/pilTools.py:7]: Why keep/use `pilImage.format`? What if the original file was in a format that PIL can read, but NSImage cannot? I'd suggest to use PNG here. There's for example no need to recompress to JPG, as NSImage will have to unpack into a pixel array anyway.

File: drawBot/drawBotDrawingTools.py [hunk 4]
```diff
@@ -394,6 +394,8 @@ def saveImage(self, path, *args, **options):
             saveImage("~/Desktop/firstImage300.png", imageResolution=300)
 
         """
+        if not isinstance(path, (str, os.PathLike)):
+            raise TypeError("Cannot apply saveImage options to multiple output formats, expected 'str', got '%s'" % type(path).__name__)
```
Comment (justvanrossum) [inline drawBot/drawBotDrawingTools.py:398]:
```suggestion
            raise TypeError("Cannot apply saveImage options to multiple output formats, expected 'str' or 'os.PathLike', got '%s'" % type(path).__name__)
```

File: tests/testExport.py [hunk 5]
```diff
@@ -210,17 +211,12 @@ def test_saveImage_unknownContext(self):
...
-                drawBot.saveImage(["foo.abcde"])
-        self.assertEqual(output.lines(), ['*** DrawBot warning: saveImage([path, path, ...]) is deprecated, use multiple saveImage statements. ***'])
-        self.assertEqual(cm.exception.args[0], "Could not find a supported context for: 'abcde'")
 
     def test_saveImage_pathList(self):
         self.makeTestDrawing()
         with self.assertRaises(TypeError) as cm:
             drawBot.saveImage(["foo.abcde"], foo=123)
-        self.assertEqual(cm.exception.args[0], 'Cannot apply saveImage options to multiple output formats.')
+        self.assertEqual(cm.exception.args[0], "Cannot apply saveImage options to multiple output formats, expected 'str', got 'list'")
```
Comment (justvanrossum) [inline tests/testExport.py:219]:
```suggestion
        self.assertEqual(cm.exception.args[0], "Cannot apply saveImage options to multiple output formats, expected 'str' or 'os.PathLike', got 'list'")
```

**Review by justvanrossum — CHANGES_REQUESTED** [review]

File: drawBot/context/gifContext.py (hunk 2 above)
Comment (typemytype) [reply drawBot/context/gifContext.py:48]: the old style `super(class, self)` happens in more places, will change this one, but the rest has to go in a separate PR

#453

File: drawBot/context/tools/pilTools.py (hunk 3 above)
Comment (typemytype) [reply drawBot/context/tools/pilTools.py:7]: +1 for PNG

Comment (typemytype) [reply drawBot/context/tools/pilTools.py:7]: now internally the image objects are build with from the TIFF representation

Comment (typemytype) [reply drawBot/context/tools/pilTools.py]: see https://github.com/typemytype/drawbot/pull/452/files#diff-7c0b8578fa67d5d3f0a2ea4b08902cf57175c06b1b8fd6b3cf0c275d804417faR13

Comment (typemytype) [reply drawBot/context/tools/pilTools.py]: an maybe `pilToNSImage`  has to move to `drawBot.misc` which is a bit more accessible then `drawBot.context.tools.pilTools`

File: drawBot/context/imageObjectContext.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import io
+import AppKit
+
+from .imageContext import ImageContext
+
+class BaseImageObjectContext(ImageContext):
+
+    def _writeDataToFile(self, data, path, options):
+        self._imageObjects = []
+        with tempfile.NamedTemporaryFile(suffix=".tiff") as temp:
```
Comment (justvanrossum) [inline drawBot/context/imageObjectContext.py:13]: Is there any way we can do thins without a temp file?

Comment (typemytype) [reply drawBot/context/imageObjectContext.py:13]: possible, now nothing is written to disk, let me check

File: drawBot/context/imageObjectContext.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+from .imageContext import ImageContext
+
+class BaseImageObjectContext(ImageContext):
+
+    def _writeDataToFile(self, data, path, options):
+        self._imageObjects = []
+        # we just need a path with a file extension
+        # nothing will be written to disk
+        super()._writeDataToFile(data, "temp.tiff", options)
```
Comment (justvanrossum) [inline drawBot/context/imageObjectContext.py:13]: Are you sure it doesn't now write to a file named "temp.tiff"?

Comment (typemytype) [reply drawBot/context/imageObjectContext.py:13]: separating writing outside the big loop helps to reuse everything from the ImageContext class

Comment (justvanrossum) [reply drawBot/context/imageObjectContext.py:13]: But now there are to methods with "ToFile" in their names which may not actually write "to file". Perhaps something for a later refactoring. I appreciate the code reuse, but it's not very clear what's going on.

Comment (typemytype) [reply drawBot/context/imageObjectContext.py:13]: true, what about removing `toFile`:

_saveImageDataToFile --> _storeImageData, _saveImageData

Comment (typemytype) [reply drawBot/context/imageObjectContext.py:13]: method renamed to _storeImageData

File: drawBot/context/gifContext.py [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import Quartz
-
+import os
+import glob
```
Comment (justvanrossum) [inline drawBot/context/gifContext.py:3]: `os` and `glob` seem unused.

**Review by justvanrossum — APPROVED** [review]

**Comment by justvanrossum** [comment]
Is this one ready to go?

**Comment by typemytype** [comment]
yep go!

**Pull request merged**